ansible_collections/arista/avd/roles/eos_designs/python_modules/ip_addressing/__init__.py

```diff
@@ -0,0 +1,328 @@
+        pool_network = ipaddress.ip_network(pool, strict=False)
+        prefixlen_diff = prefixlen - pool_network.prefixlen
+        subnet_size = (int(pool_network.hostmask) + 1) >> prefixlen_diff
+        if (subnet_offset + 1) * subnet_size > pool_network.num_addresses:
+            raise AristaAvdError(f"Unable to get {subnet_offset + 1} /{prefixlen} subnets from pool {pool}")
+        subnet = ipaddress.ip_network((int(pool_network.network_address) + subnet_offset * subnet_size, prefixlen))
+        try:
+            ip = subnet[ip_offset]
+        except IndexError as e:
+            raise AristaAvdError(f"Unable to get {ip_offset+1} hosts in subnet {subnet} taken from pool {pool}") from e
+        return str(ip)
+
+    def _template(self, template_path, **kwargs):
+        template_vars = ChainMap(kwargs, self._hostvars)
+        return self.template_var(template_path, template_vars)
+
+    @cached_property
+    def _mlag_primary_id(self) -> int:
+        return int(get(self._hostvars, "switch.mlag_switch_ids.primary", required=True))
+
+    @cached_property
+    def _mlag_secondary_id(self) -> int:
+        return int(get(self._hostvars, "switch.mlag_switch_ids.secondary", required=True))
+
+    @cached_property
+    def _mlag_peer_ipv4_pool(self) -> str:
+        return get(self._hostvars, "switch.mlag_peer_ipv4_pool", required=True)
+
+    @cached_property
+    def _mlag_peer_l3_ipv4_pool(self) -> str:
+        return get(self._hostvars, "switch.mlag_peer_l3_ipv4_pool", required=True)
+
+    @cached_property
+    def _uplink_ipv4_pool(self) -> str:
+        return get(self._hostvars, "switch.uplink_ipv4_pool", required=True)
+
+    @cached_property
+    def _id(self) -> int:
+        return int(get(self._hostvars, "switch.id", required=True))
+
+    @cached_property
+    def _max_uplink_switches(self) -> int:
+        return int(get(self._hostvars, "switch.max_uplink_switches", required=True))
+
+    @cached_property
+    def _max_parallel_uplinks(self) -> int:
+        return int(get(self._hostvars, "switch.max_parallel_uplinks", required=True))
+
+    @cached_property
+    def _loopback_ipv4_pool(self) -> str:
+        return get(self._hostvars, "switch.loopback_ipv4_pool", required=True)
+
+    @cached_property
+    def _loopback_ipv4_offset(self) -> int:
+        return get(self._hostvars, "switch.loopback_ipv4_offset", required=True)
+
+    @cached_property
+    def _loopback_ipv6_pool(self) -> str:
+        return get(self._hostvars, "switch.loopback_ipv6_pool", required=True)
+
+    @cached_property
+    def _loopback_ipv6_offset(self) -> int:
+        return get(self._hostvars, "switch.loopback_ipv6_offset", required=True)
+
+    @cached_property
+    def _vtep_loopback_ipv4_pool(self) -> str:
+        return get(self._hostvars, "switch.vtep_loopback_ipv4_pool", required=True)
+
+    def mlag_ibgp_peering_ip_primary(self, mlag_ibgp_peering_ipv4_pool: str) -> str:
```

Comment:
So given that these are almost the same function rather than duplicating code I think it could be better to just add to these pairs of functions an input argutment being "role" (primary or secondary) and to return accordingly one or the other

This would reduce the code here as well as reducing the code in eos_designs_facts.py

```
if self.mlag_role == "primary":
  return self.avd_ip_address.mlag_ibgp_peering_ip_primary()
elif self.mlag_role == "secondary":
  return self.avd_ip_address.mlag_ibgp_peering_ip_secondary()
```

could just become
```
  return self.avd_ip_address.mlag_ibgp_peering_ip_primary(self.mlag_role)
```


The only thing it would really change is that now people would need to implement method with similar signatures when injecting their own.

Reply:
We need to respect the existing split of templates. Otherwise it may not work with existing deployments using custom templates.

Reply:
if it was a string then yes. But it is a function call.

Reply:
We agreed to keep it as is for now and we will revisit later.